### ISSUE TITLE
Toolchain probing cannot cope with warnings from java

### ISSUE THREAD

**Issue by jskov-jyskebank-dk** [gradle/gradle#21616]
### Expected Behavior

Gradle should be able to handle "java -version" output that contains warnings:

```console
$ java -version
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
openjdk version "17.0.4" 2022-07-19 LTS
OpenJDK Runtime Environment (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS, mixed mode, sharing)
```

### Current Behavior

Gradle (7.5.1) `javaToolchains` task fails with:

```
(14:47:50)  + Options
(14:47:50)      | Auto-detection:     Enabled
(14:47:50)      | Auto-download:      Enabled
(14:47:50) 
(14:47:50)  + Red Hat, Inc. JDK 17.0.4+8-LTS
(14:47:50)      | Location:           /usr/lib/jvm/java-17-openjdk-17.0.4.0.8-2.el8_6.x86_64
(14:47:50)      | Language Version:   17
(14:47:50)      | Vendor:             Red Hat, Inc.
(14:47:50)      | Architecture:       amd64
(14:47:50)      | Is JDK:             true
(14:47:50)      | Detected by:        Common Linux Locations
(14:47:50) 
(14:47:50)  + Invalid toolchains
(14:47:50)      + /usr/lib/jvm/java-11-openjdk-11.0.16.0.8-1.el8_6.x86_64
(14:47:50)        | Error:              Unexpected command output: 
(14:47:50) [0.015s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
(14:47:50) [0.015s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
(14:47:50) [0.015s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
(14:47:50) /usr/lib/jvm/java-11-openjdk-11.0.16.0.8-1.el8_6.x86_64
(14:47:50) 11.0.16
(14:47:50) Red Hat, Inc.
(14:47:50) amd64
(14:47:50) OpenJDK 64-Bit Server VM
(14:47:50) 11.0.16+8-LTS
(14:47:50) OpenJDK Runtime Environment
(14:47:50) 11.0.16+8-LTS
```

### Context

This happens when we run Podman containers on OpenShift - Podman runs as non-root. And Gradle inside of this.
So this is a two-layer deep container context (on top of core OS).

Lots of interesting OCP/Podman/Kernel magic is in play that I cannot possibly describe fully.
But I suspect that the JDK warns about the fact that it runs in a container - or some aspect of this, which it is surprised about.

### Steps to Reproduce 

It is not easy to reproduce (as per Context). But I will be super happy to help test a fix!

I can provide you with this output, which I hope is enough for you to accept the issue.

The output is from running `java -version` (showing what is stdout and what is stderr):

```
STEP 2/4: RUN java -version
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.056s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
openjdk version "17.0.4" 2022-07-19 LTS
OpenJDK Runtime Environment (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS, mixed mode, sharing)
--> c52c806705f
STEP 3/4: RUN java -version >/dev/null
openjdk version "17.0.4" 2022-07-19 LTS
OpenJDK Runtime Environment (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.4.0.8-2.el9_0) (build 17.0.4+8-LTS, mixed mode, sharing)
--> 955548f0dcc
STEP 4/4: RUN java -version 2>/dev/null
[0.004s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.004s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
[0.004s][warning][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /sys/fs/cgroup/cpuset.
```

### Other hints

The only other (relevant) reference to this problem with java that I could find is: https://github.com/NixOS/nixpkgs/issues/163164#issuecomment-1122582269

Without knowing all the intricacies of parsing output from all the possible variants and versions of java, I would suggest using only the stderr output from 'java -version'. But that may of course break other stuff.
Or maybe strip out lines beginning with [ if that is a standard for warnings used in all java variants/versions.

**Comment by jbartok** [comment]
I can confirm that this problem happens due to the interaction of two things:

1. the JVM in question, in the environment it's running in, prints some warning messages to standard output (the JVM running, in this case, is a process started by Gradle in order to [probe a Java installation, by running its `bin/java`](https://github.com/gradle/gradle/blob/7f8686925b7299e6350b96b1d90ac9bb4ae1d00c/subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/DefaultJvmMetadataDetector.java#L98) to figure out the installation's metadata, ie. system properties)
2. the [class executed by said JVM prints system properties to standard output](https://github.com/gradle/gradle/blob/7f8686925b7299e6350b96b1d90ac9bb4ae1d00c/subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/MetadataProbe.java#L43) (to be examined by Gradle at a later stage), in an overly simplistic manner, ie. just a list of property values in a fixed order

Due to the simplistic output the probing process uses, it can't handle extra lines, the JVM warnings in this case.

We could fix it by, for example using a key-value pair type of output, which then could be filtered for valid lines. I will leave this issue open, maybe we'll get to do this at some point, but honestly, considering our resource limitations, I wouldn't hold my breath...

As a workaround, maybe you can set the `PrintWarnings` JVM property to `false`, to get rid of those warning.

**Comment by jbartok** [comment]
Actually, you don't have a way of getting that JVM property to the right process... Will try to find another way.

**Comment by alexander-schubert** [comment]
The following work-around seem to be viable for Java >=9 (`-Xlog` was only introduced in Java 9, but the issue currently does not seem to exist on older versions).

**Work-around**
Set `JAVA_TOOL_OPTIONS="-Xlog:disable -Xlog:all=warning:stderr:uptime,level,tags"` in the environment where you're executing Gradle to redirect all warnings to stderr instead.

**Limitations**
Note that this also affects other `java` processes and prints an ugly `Picked up JAVA_TOOL_OPTIONS` message for every process launch. Also, likely due some caching in Gradle, you may need to restart your Gradle daemon (`gradle --stop`) for the work-around to be effective.

**Comment by jbartok** [comment]
@jskov-jyskebank-dk I've merged in a change which should make the probing capable of ignoring the extra lines. Should be in **tomorrow's** nightly build (https://gradle.org/release-nightly/); pls. try it out, if you can

**Comment by jskov-jyskebank-dk** [comment]
Will do, thanks!

**Comment by jbartok** [comment]
@jskov-jyskebank-dk, I will close this issue now, pls. re-open it if the fix doesn't solve your problem or anything else is wrong with it.

**Comment by jskov-jyskebank-dk** [comment]
@jbartok noted.
I did not find time (yet), but will get around to it eventually. And I will report success/failure here for completeness.
Cheers!

**Comment by jskov-jyskebank-dk** [comment]
I can (finally) confirm that the fix works for me.

Thanks @jbartok !

**Comment by pantherdd** [comment]
Seems like the same warnings were causing #20837 too, can someone confirm?

**Comment by jbartok** [comment]
@pantherdd no, #20837 is unrelated and will be fixed separately.

**Comment by pantherdd** [comment]
I see, it's probably just a coincidence then. I thought these may be related since we faced that same `Error while evaluating property 'javaVersion'` exception I mentioned above when setting `options.forkOptions.javaHome` to the absolute path of a JDK that wrote similar warnings to the console when running `${javaHome}/bin/java -version`. When I changed `javaHome` to a JDK that did not output such warnings, the issue went away.

**Comment by pantherdd** [comment]
For those led here by Google, I just checked, and the issue I had (quoted below) is now fixed in Gradle 7.6.

> Seems like the same `java -version` warnings were causing the `Error while evaluating property 'javaVersion' of task ':compileJava'` issues too (e.g. #20837), can someone confirm?
> 
> ```
> org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileJava'.
>         at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:38)
>         at ...
> Caused by: org.gradle.api.internal.tasks.properties.PropertyEvaluationException: Error while evaluating property 'javaVersion' of task ':compileJava'
>         at org.gradle.api.internal.tasks.properties.InputParameterUtils.prepareInputParameterValue(InputParameterUtils.java:33)
>         at ...
> Caused by: java.util.NoSuchElementException: No value present
>         at org.gradle.jvm.toolchain.internal.JavaToolchainQueryService.query(JavaToolchainQueryService.java:74)
>         at ...
> ```
